### PR TITLE
Minor tweaks to handling of PHA settings

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Improve the checking of the factor and rate settings of set_analysis and the plot_fac attribute of the DataPHA class, and add a few tests that exercise some corner cases of the DataPHA class.

# Details

The plot_fac value was being silently converted to an integer - so `pha.plot_fac = 0.5` or `set_analysis("energy", factor=0.5)` would set `plot_fac` to 0. We now error out it a non-integer-like value is used (so 2.0 is okay, as it was before, but "`" is no-longer supported). I would rather error out than make the code behave "strangely" for the user. An alternative would be to just warn when the value has been truncated, but I am guessing that this is not-likely to be happening in user code.

The type argument to set_analysis was strange - it would allow any string which, when converted to lower-case, began with "rate" or "counts" - **but** would only set the rate option if `str(type).lower() == "rate"`. This is not a good UX. So we now require the value to be either "counts" or "rate" (and no case conversion or stripping of spaces). The existing documentation lists these as the only two valid arguments.

I threw in some minor "style" changes for `get_areascal/get_backscal`.

As well as testing these changes I've added a few other tests to catch DataPHA code that we weren't testing (edge cases).

There are a few style changes to the tests

- we can now safely import sherpa.astro.io thanks to #1185, and some "standardisation" of checks on the backend - maybe something we want to add to `conftest.py` in a later PR related to #1362
- use of pytest.approx rather than numpy.assert_allclose
- swapping from ' to " to mark a string, following the black standards, to see how this works, related to #1369

### CODE REVIEW

File: sherpa/astro/data.py [hunk 1]
```diff
@@ -1523,23 +1523,43 @@ def _get_rate(self):
...
+            okay = val == int(val)
+        except (ValueError, TypeError):
+            # For when int() can't convert val, which can raise
+            # different errors.
+            okay = False
+
+        if not okay:
+            raise DataErr("bad", "plot_fac setting", val)
+
+        val = int(val)
```
Comment (hamogu) [inline sherpa/astro/data.py:1552]:
```suggestion
        try:
            val == int(val)
        except (ValueError, TypeError):
            # For when int() can't convert val, which can raise
            # different errors.
            raise DataErr("bad", "plot_fac setting", val)
```
I don't understand that extra "okay" variable. If `val = int(val)` works, then val is an int and we don't need okey. If it dies not, then we raise an Error anyway.

File: sherpa/astro/data.py [hunk 2]
```diff
@@ -1523,23 +1523,43 @@ def _get_rate(self):
...
         for id in self.background_ids:
             self.get_background(id).plot_fac = val
 
     plot_fac = property(_get_plot_fac, _set_plot_fac,
-                        doc='Number of times to multiply the y-axis ' +
-                        'quantity by x-axis bin size')
+                        doc="""How the X axis is used to create the Y axis when plotting data.
+
+The Y axis values are multiplied by X^plot_fac. The default
+value of 0 means the X axis is not used in plots. The value
```
Comment (hamogu) [inline sherpa/astro/data.py:1561]: That means the plots don't have an x-axis?

File: sherpa/astro/tests/test_astro_data2.py [hunk 3]
```diff
@@ -1111,6 +1111,58 @@ def test_pha_invalid_units(invalid, make_test_pha):
     assert str(de.value) == f"unknown quantity: '{invalid}'"
 
 
+# currently several variants are allowed, even if they don't work as
+# expected - e.g. type="rates" does not set the rate value to True.
```
Comment (hamogu) [inline sherpa/astro/tests/test_astro_data2.py:1115]: Wait, I thought this PR fixes that so that the variants that don't work as expected no longer work?

**Review by hamogu — CHANGES_REQUESTED** [review]

File: sherpa/astro/data.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/data.py:1552]: So, this bit of code went through several revisions, so maybe I just got confused. What is was worried about for your version was that the val that would get used in the `raise` call to create the error message would have already been converted to an integer, and so the message would have been confusing to the user. But I now realize that that thinking is baloney, so I'll make the change.

File: sherpa/astro/data.py (hunk 2 above)
Comment (DougBurke) [reply sherpa/astro/data.py:1561]: I was trying to come up with something meaningful to say here, but failed. The problem is that I don't really want to say "the y axis is multiplied by x axis^plot_fac" (even though I do actually say this) as it is logically wrong - the y axis is the y axis. What I'm really trying to say is that the "base units" are multiplied by x^plot_fac values to create the y axis, but there's no definition of "base units" (it depends on the rate setting, for one thing, and the plot type for another). So options include:

- dropping this text, just leaving the first line
- trying to come up with some very basic text that doesn't really say anything (and if so, is it worth keeping)
- coming up with a better way to describe this

Any ideas?

File: sherpa/astro/tests/test_astro_data2.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/tests/test_astro_data2.py:1115]: Did I mess something up in my rebasing? These tests shouldn't be marked as xfail (and the comment above should be removed). Let me check

File: sherpa/astro/data.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/data.py:1552]: Actually, no. I think I do need the existing code. I want to allow

    factor=1.0

(since this was allowed before, so I don't see it worth breaking code) but disallow

    factor = 1.5

(since this leads to potentially-surprising results) and so you need the current version - or perhaps better would be

```
    try:
        ival = int(val)
        okay = val == ival
    except (...):
        okay = False

    if okay:
        raise ...
```

An alternative to error-ing out would be to create a warning message, but my guess is that not many people have used non-integer-like values (and  that this is technically different to allowing 1.0).

File: sherpa/astro/tests/test_astro_data2.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/tests/test_astro_data2.py]: My rebase left these in when they can be removed (ie the xfail tests suddenly started passsing). So I've made the change in "remove un-needed xfail markers".

File: sherpa/astro/data.py [hunk 4]
```diff
@@ -1523,23 +1523,44 @@ def _get_rate(self):
...
+        try:
+            ival = int(val)
+            okay = val == ival
+        except (ValueError, TypeError):
+            # For when int() can't convert val, which can raise
+            # different errors.
+            okay = False
+
+        if not okay:
+            raise DataErr("bad", "plot_fac setting", val)
```
Comment (hamogu) [inline sherpa/astro/data.py:1552]: I think I still see a problem here, but I need to play around with the old version more to see what the differences are. Unfortunately, I might not get to that today.

Comment (DougBurke) [reply sherpa/astro/data.py:1552]: There's no hurry.

Comment (DougBurke) [reply sherpa/astro/data.py:1552]: If we didn't want to try and be backwards compatible - i.e. we don't need to support "set_analysis(... , factor=2.0)" - then it would be easier. I note in matplotlib 5.3.0 that they have a bunch of changes like this (I think), but I'm not sure our user community has the same setup as they do.

I'm not particularly happy with the code, but it works (as far as I can tell).

File: sherpa/astro/data.py (hunk 2 above)
Comment (hamogu) [reply sherpa/astro/data.py:1562]: How about:
```
'''Scale y values when plotting data

y-axis values can be multiplied by x^plot_fac
where x is the corresponding x-value for each datapoint. 
The default is plot_fac=0, so no scaling is applied. This is useful to show e.g. counts/kev on the y-axis of the plot.
'''
```
(Does my example make any sense?)

File: sherpa/astro/data.py (hunk 4 above)
Comment (hamogu) [reply sherpa/astro/data.py:1552]: What exactly is this supposed to accept?
- Do inputs have to be integers, i.e. is there a good physical reason to allow `factor=2`, but not `factor=3/2`? That asks for what scalings make sense. Surely 0 and 1 do make sense, but why not 0.5? I can imagine someone scaling with sqrt(E) or sqrt(lambda), in which case such a plot would make sense or not?
- What inputs do we want to accept? Int, float, also string? I can only trigger the except clause with exotic cases like strings, complex number etc. If the goal of the except clause is to give a better error message for strings, etc., I would suggest:
```
except (ValueError, TypeError):
    raise ArgumentTypeErr("badarg", val, "int or float")
```

Code that I used to try to understand what happens
```
def _set_plot_fac(val):
    '''removed "self" argument and turned output and error into print statements
    so I can call it a couple of times in a row'''
    try:
        ival = int(val)
        okay = val == ival
    except (ValueError, TypeError):
        # For when int() can't convert val, which can raise
        # different errors.
        print("except clause triggered")
        okay = False

    if not okay:
        print(f"{val=} not okay")
        return
            
    print(ival)


0
1
2
val=2.5 not okay
0
1
2
val=2.5 not okay
val='0' not okay
val='1' not okay
val='2' not okay
except clause triggered
val='2.5' not okay
except clause triggered
val='0.' not okay
except clause triggered
val='1.' not okay
except clause triggered
val='2.' not okay
except clause triggered
val='2.5' not okay
```

Comment (DougBurke) [reply sherpa/astro/data.py:1552]: It accepts an integer. There is no support for non-integer values. adding support for factor=0.5 would be a much-larger PR. This is only about improving the existing support.

When setting plot_fac - either directly as an attribute or via the factor argument of set_analysis - the original code would take the value you send it (`val`) and then call int on it (`int(val)`) - so that if you sent it 0.5 it would silelntly get converted to 0. All I'm really trying to do is ensure we test these assumptions.

So your output above looks like I'd expect it to.

The one question is that you used to be able to set plot_fac to "1" or "2" and this is no-longer supported, but I am pretty sure we never documented support for this. There are a number of places in our code base hwere we can accept strings rather than numbers which comes from "pre history" (e.g. when supporting S-Lang as well as Python and a spill over from the old command language for Sherpa).

**Review by hamogu — APPROVED** [review]

File: sherpa/astro/data.py (hunk 2 above)
Comment (hamogu) [reply sherpa/astro/data.py:1562]: I've approved it already. I don't think my wording is that much better than what's in the PR now, but feel free to use either or a combination of both.

**Comment by DougBurke** [comment]
I've rebased this to address a merge conflict / slight-change-in-behavior. In doing so I took advantage of moving some code around to remove one of the commits.
